### PR TITLE
Added new boundary tests & fixed boundary check bug in hwloop & mac

### PR DESCRIPTION
**Pull request by jessicamills**
Also: 1) clean up a couple of inaccurate comments in two earlier HW Loop
            tests; and 2) change an existing MAC test to use a non-boundary value.

  gas/ChangeLog.COREV:
    
            * config/tc-riscv.c: Fixed bug in hardware loop operand
            boundary checks (b1 and b2).
    
    gas/testsuite/ChangeLog.COREV:
    
            * cv-hwloop-16.d: Added new boundary test.
            * cv-hwloop-16.l: Likewise.
            * cv-hwloop-16.s: Likewise.
            * cv-hwloop-17.d: Likewise.
            * cv-hwloop-17.l: Likewise.
            * cv-hwloop-17.s: Likewise.
            * cv-hwloop-18.d: Likewise.
            * cv-hwloop-18.l: Likewise.
            * cv-hwloop-18.s: Likewise.
            * cv-hwloop-19.d: Likewise.
            * cv-hwloop-19.l: Likewise.
            * cv-hwloop-19.s: Likewise.
            * cv-hwloop-20.d: Likewise.
            * cv-hwloop-20.l: Likewise.
            * cv-hwloop-20.s: Likewise.
            * cv-hwloop-21.d: Likewise.
            * cv-hwloop-21.l: Likewise.
            * cv-hwloop-21.s: Likewise.
            * cv-hwloop-22.d: Likewise.
            * cv-hwloop-22.l: Likewise.
            * cv-hwloop-22.s: Likewise.
            * cv-hwloop-23.d: Likewise.
            * cv-hwloop-23.l: Likewise.
            * cv-hwloop-23.s: Likewise.
            * cv-hwloop-09.s: Fixed incorrect comment.
            * cv-hwloop-10.s: Likewise.
            * cv-mac-07.d: Added new boundary test.
            * cv-mac-07.l: Likewise.
            * cv-mac-07.s: Likewise.
            * cv-mac-08.d: Likewise.
            * cv-mac-08.l: Likewise.
            * cv-mac-08.s: Likewise.
            * cv-mac-05.l: Changed to non-boundary value.
            * cv-mac-05.s: Likewise.
    
    Signed-off-by: Jessica Mills <jessica.mills@embecosm.com>

### CODE REVIEW

File: gas/testsuite/gas/riscv/cv-hwloop-23.s [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Loop count must be positive integer in range:[0, 4094]
+target:
+	cv.setupi 0, 4095, 4
```
Comment (flip1995) [inline gas/testsuite/gas/riscv/cv-hwloop-23.s]: Something's wrong with this test. The second operand of `cv.setupi` is `uimmS` (5-bit imm) and should be in the range `[0,30]`.

Did you mean to write
```suggestion
# Loop count must be positive integer in range:[0, 4094]
target:
	cv.setupi 0, 4, 4095
```

?

This still does not emit the above error message, though. The 3rd operand in `cv.setupi` has not to be aligned on a (half) word boundary.

If the error message, you expect in `gas/testsuite/gas/riscv/cv-hwloop-23.l` should be emitted, this test has to be
```suggestion
# Loop count must be positive integer in range:[0, 4095]
target:
	cv.setupi 0, 4, 4096
```

**Review by flip1995 — CHANGES_REQUESTED** [review]
After reviewing this, I'm confused. Is the documentation wrong, am I misunderstanding it or is your implementation wrong?
image


In the "Short Hardware Loop Setup Instructions" section, the `setupi` description has the operands in the order `L, uimmL, uimmS`. The "Encoding" part has it in the order `lp.setupi L, uimmS, uimmL`. I assumed, that the documented Mnemonic is correct and implemented it like this in LLVM. So... WDYT?

File: gas/testsuite/gas/riscv/cv-hwloop-23.s (hunk 1 above)
Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: @flip1995 The right mneumonic for cv.setupi is L, uimmL, uimmS. These are good questions to ask us as we will have come across this before. The documentation is often incorrect, when you come across these the best thing to do is check the the pulp binutils-gdb to see how it is implemented.

Comment (flip1995) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: Ah, that's good to know. Then I will fix this in LLVM.

Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: @flip1995 Following on from this, I think the tests are fine. Please let me know if you have any further questions :)

Comment (flip1995) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: Something with the test is still wrong though. I guess you meant to test `4096 ∉ [0, 4095]`? The value `4095` is fine in the `uimmL` position of `cv.setupi`.

Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: @flip1995 Ahh yes, good spotting, will change this!

Comment (jessicamills) [reply gas/testsuite/gas/riscv/cv-hwloop-23.s]: Just fixed the error - this was down to an incorrect comment. Thanks @flip1995

File: gas/testsuite/ChangeLog.COREV [hunk 2]
```diff
@@ -1,3 +1,38 @@
+2020-11-19  Jessica Mills  <jessica.mills@embecosm.com>
```
Comment (jeremybennett) [inline gas/testsuite/ChangeLog.COREV:1]: You only list the new files, not the files you have changed (e.g. `cv-hwloop-09.s`)

**Review by jeremybennett — CHANGES_REQUESTED** [review]
This is really two issues being fixed.

1. You are adding some new boundary tests

2. You are modifying some existing tests to correct boundary values.

The ChangeLog only mentions the first of these and needs correcting.

A general idea is that one commit should deal with one issue, so I suggest splitting this into two pull requests, one to fix the existing tests, the second to add the new boundary tests.

**Comment by jessicamills** [comment]
@jeremybennett None of the previous tests are corrected in this pull request - the only issue with these is the comment was wrong. This incorrect comment lead to me putting in the wrong value for a new test. I have added in the files with the changed comment to the ChangeLog. If you think the changed comment requires a different request I am happy to do this.

File: gas/testsuite/ChangeLog.COREV (hunk 2 above)
Comment (jeremybennett) [reply gas/testsuite/ChangeLog.COREV:1]: Now corrected

**Comment by jeremybennett** [comment]
OK - I had failed to appreciate it was just cleaning up a couple of comments and tidying up one test.

It is too much hassle to make this a separate commit.  I would deal with this by adding a paragraph after the first line of the commit, above the ChangeLog entries. I've made this change for you.

**Review by jeremybennett — APPROVED** [review]
